### ISSUE TITLE
Explicitly ignoring OK/passing validation results does not work

### ISSUE THREAD

**Issue by atc0005** [atc0005/check-cert#525]
## Overview

~Need to audit the logic for ignoring OK results and make sure that it is working as intended for all validation check types.~ Even if a validation check is successful/OK, if the sysadmin requested that the result be ignored we need to honor that. Brief testing seems to confirm that this is not working as intended.

In particular, ignoring (successful) hostname validation results does not seem to work properly.

## References

- https://github.com/atc0005/check-cert/issues/505

**Comment by atc0005** [comment]
Results from checking www.google.com:

```console
$ go build ./cmd/check_cert && ./check_cert.exe --server 'www.google.com' --ignore-validation-result 'expiration,hostname' --log-level debug
8:00AM DBG cmd\check_cert\main.go:147 > Expanding given host pattern in order to obtain IP Address age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG cmd\check_cert\main.go:220 > Multiple IP Addresses resolved from given host pattern age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= ip_addresses=["108.177.122.106","108.177.122.147","108.177.122.99","108.177.122.103","108.177.122.104","108.177.122.105"] logging_level=debug num_ip_addresses=6 port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG cmd\check_cert\main.go:221 > Using first IP Address, ignoring others age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG cmd\check_cert\main.go:294 > Retrieving certificate chain age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s dns_name= expected_sans_entries= filename= host_value=www.google.com ip_address=108.177.122.106 logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG internal\netutils\net.go:249 > Connecting to remote server age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= host=www.google.com ip_address=108.177.122.106 logging_level=debug port=443 server=www.google.com timeout=10s version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG internal\netutils\net.go:287 > Connected age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= host=www.google.com ip_address=108.177.122.106 logging_level=debug port=443 server=www.google.com timeout=10s version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG internal\netutils\net.go:291 > Retrieved certificate chain age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= host=www.google.com ip_address=108.177.122.106 logging_level=debug port=443 server=www.google.com timeout=10s version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG internal\netutils\net.go:300 > Successfully closed connection to server age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= host=www.google.com ip_address=108.177.122.106 logging_level=debug port=443 server=www.google.com timeout=10s
 version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG cmd\check_cert\validate.go:50 > Hostname validation successful age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG cmd\check_cert\validate.go:74 > SANs List validation ignored age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG cmd\check_cert\validate.go:110 > Expiration validation ignored age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com total_certificates=3 version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG cmd\check_cert\main.go:378 > Name: Hostname, Priority: 2
 age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG cmd\check_cert\main.go:378 > Name: SANs List, Priority: 1
 age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG cmd\check_cert\main.go:378 > Name: Expiration, Priority: 3
 age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG cmd\check_cert\main.go:441 > No (non-ignored) problems with certificate chain detected age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s checks_failed=0 checks_ignored=2 checks_successful=1 checks_total=3 expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:00AM DBG cmd\check_cert\annotations.go:46 > It took 0s to execute annotateError func (errors evaluated: 0, annotated: 0) age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=false apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
OK: Hostname validation using value "www.google.com" successful for leaf certificate [checks: 2 IGNORED (SANs List, Expiration), 0 FAILED, 1 SUCCESSFUL (Hostname)]

3 certs retrieved for service running on www.google.com (108.177.122.106) at port 443 using host value "www.google.com"


PROBLEM RESULTS:

* None


IGNORED RESULTS:

[--] Expiration validation ignored: 0 expired certificates, 0 expiring certificates

Certificate 1 of 3 (leaf):
        Name: CN=www.google.com
        SANs entries: [www.google.com]
        Issuer: CN=GTS CA 1C3,O=Google Trust Services LLC,C=US
        Serial: 63:4E:19:CA:90:36:79:74:0A:B2:94:4F:E8:A3:E7:C9
        Issued On: 2023-02-08 04:36:30 +0000 UTC
        Expiration: 2023-05-03 04:36:29 +0000 UTC
        Status: [OK] 63d 14h remaining

Certificate 2 of 3 (intermediate):
        Name: CN=GTS CA 1C3,O=Google Trust Services LLC,C=US
        SANs entries: []
        Issuer: CN=GTS Root R1,O=Google Trust Services LLC,C=US
        Serial: 02:03:BC:53:59:6B:34:C7:18:F5:01:50:66
        Issued On: 2020-08-13 00:00:42 +0000 UTC
        Expiration: 2027-09-30 00:00:42 +0000 UTC
        Status: [OK] 1674d 10h remaining

Certificate 3 of 3 (intermediate):
        Name: CN=GTS Root R1,O=Google Trust Services LLC,C=US
        SANs entries: []
        Issuer: CN=GlobalSign Root CA,OU=Root CA,O=GlobalSign nv-sa,C=BE
        Serial: 77:BD:0D:6C:DB:36:F9:1A:EA:21:0F:C4:F0:58:D3:0D
        Issued On: 2020-06-19 00:00:42 +0000 UTC
        Expiration: 2028-01-28 00:00:42 +0000 UTC
        Status: [OK] 1794d 10h remaining

[--] SANs List validation ignored: 0 SANs entries specified, 1 SANs entries on leaf cert [0 EXPECTED, 0 MISSING, 0 UNEXPECTED]


SUCCESS RESULTS:

[OK] Hostname validation using value "www.google.com" successful for leaf certificate

 | 'certs_present_intermediate'=2;;;; 'certs_present_leaf'=1;;;; 'certs_present_root'=0;;;; 'certs_present_unknown'=0;;;; 'expires_intermediate'=1674d;30;15;0; 'expires_leaf'=63d;30;15;0; 'time'=51ms;;;;
```

```console
$ go build ./cmd/check_cert && ./check_cert.exe --server 'www.google.com' --ignore-validation-result 'hostname' --log-level debug
8:01AM DBG cmd\check_cert\main.go:147 > Expanding given host pattern in order to obtain IP Address age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG cmd\check_cert\main.go:220 > Multiple IP Addresses resolved from given host pattern age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= ip_addresses=["108.177.122.106","108.177.122.147","108.177.122.99","108.177.122.103","108.177.122.104","108.177.122.105"] logging_level=debug num_ip_addresses=6 port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG cmd\check_cert\main.go:221 > Using first IP Address, ignoring others age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG cmd\check_cert\main.go:294 > Retrieving certificate chain age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s dns_name= expected_sans_entries= filename= host_value=www.google.com ip_address=108.177.122.106 logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG internal\netutils\net.go:249 > Connecting to remote server age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= host=www.google.com ip_address=108.177.122.106 logging_level=debug port=443 server=www.google.com timeout=10s version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG internal\netutils\net.go:287 > Connected age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= host=www.google.com ip_address=108.177.122.106 logging_level=debug port=443 server=www.google.com timeout=10s version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG internal\netutils\net.go:291 > Retrieved certificate chain age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= host=www.google.com ip_address=108.177.122.106 logging_level=debug port=443 server=www.google.com timeout=10s version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG internal\netutils\net.go:300 > Successfully closed connection to server age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= host=www.google.com ip_address=108.177.122.106 logging_level=debug port=443 server=www.google.com timeout=10s version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG cmd\check_cert\validate.go:50 > Hostname validation successful age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG cmd\check_cert\validate.go:74 > SANs List validation ignored age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG cmd\check_cert\validate.go:118 > Expiration validation successful age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= expired_certificates=0 expiring_certificates=0 filename= logging_level=debug port=443 server=www.google.com total_certificates=3 valid_certificates=3 version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG cmd\check_cert\main.go:378 > Name: Hostname, Priority: 2
 age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG cmd\check_cert\main.go:378 > Name: SANs List, Priority: 1
 age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG cmd\check_cert\main.go:378 > Name: Expiration, Priority: 3
 age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG cmd\check_cert\main.go:441 > No (non-ignored) problems with certificate chain detected age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s checks_failed=0 checks_ignored=1 checks_successful=2 checks_total=3 expected_sans_entries= filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
8:01AM DBG cmd\check_cert\annotations.go:46 > It took 0s to execute annotateError func (errors evaluated: 0, annotated: 0) age_critical=15 age_warning=30 app_type=plugin apply_expiration_validation_results=true apply_hostname_validation_results=false apply_sans_list_validation_results=false cert_check_timeout=10s filename= logging_level=debug port=443 server=www.google.com version="check-cert x.y.z (https://github.com/atc0005/check-cert)"
OK: Expiration validation successful: leaf cert "www.google.com" expires next with 63d 14h remaining (until 2023-05-03 04:36:29 +0000 UTC) [checks: 1 IGNORED (SANs List), 0 FAILED, 2 SUCCESSFUL (Hostname, Expiration)]

3 certs retrieved for service running on www.google.com (108.177.122.106) at port 443 using host value "www.google.com"


PROBLEM RESULTS:

* None


IGNORED RESULTS:

[--] SANs List validation ignored: 0 SANs entries specified, 1 SANs entries on leaf cert [0 EXPECTED, 0 MISSING, 0 UNEXPECTED]


SUCCESS RESULTS:

[OK] Expiration validation successful: leaf cert "www.google.com" expires next with 63d 14h remaining (until 2023-05-03 04:36:29 +0000 UTC)

Certificate 1 of 3 (leaf):
        Name: CN=www.google.com
        SANs entries: [www.google.com]
        Issuer: CN=GTS CA 1C3,O=Google Trust Services LLC,C=US
        Serial: 63:4E:19:CA:90:36:79:74:0A:B2:94:4F:E8:A3:E7:C9
        Issued On: 2023-02-08 04:36:30 +0000 UTC
        Expiration: 2023-05-03 04:36:29 +0000 UTC
        Status: [OK] 63d 14h remaining

Certificate 2 of 3 (intermediate):
        Name: CN=GTS CA 1C3,O=Google Trust Services LLC,C=US
        SANs entries: []
        Issuer: CN=GTS Root R1,O=Google Trust Services LLC,C=US
        Serial: 02:03:BC:53:59:6B:34:C7:18:F5:01:50:66
        Issued On: 2020-08-13 00:00:42 +0000 UTC
        Expiration: 2027-09-30 00:00:42 +0000 UTC
        Status: [OK] 1674d 9h remaining

Certificate 3 of 3 (intermediate):
        Name: CN=GTS Root R1,O=Google Trust Services LLC,C=US
        SANs entries: []
        Issuer: CN=GlobalSign Root CA,OU=Root CA,O=GlobalSign nv-sa,C=BE
        Serial: 77:BD:0D:6C:DB:36:F9:1A:EA:21:0F:C4:F0:58:D3:0D
        Issued On: 2020-06-19 00:00:42 +0000 UTC
        Expiration: 2028-01-28 00:00:42 +0000 UTC
        Status: [OK] 1794d 9h remaining

[OK] Hostname validation using value "www.google.com" successful for leaf certificate

 | 'certs_present_intermediate'=2;;;; 'certs_present_leaf'=1;;;; 'certs_present_root'=0;;;; 'certs_present_unknown'=0;;;; 'expires_intermediate'=1674d;30;15;0; 'expires_leaf'=63d;30;15;0; 'time'=27ms;;;;
```

Looks like an issue with ignoring a successful hostname validation check result remains.

**Comment by atc0005** [comment]
> In particular, ignoring (successful) hostname validation results does not seem to work properly.

Fixed per #528.

As of 1600db2a7f372d227fcda56b5a6b7f469f8e3bdf, this is the result from disabling all validation checks:

```console
$ ./check_cert.exe --server www.google.com --ignore-validation-result hostname,expiration,sans
OK: Expiration validation ignored: leaf cert "www.google.com" expires next with 63d 7h remaining (until 2023-05-03 04:36:29 +0000 UTC) [checks: 3 IGNORED (Expiration, Hostname, SANs List), 0 FAILED, 0 SUCCESSFUL]

3 certs retrieved for service running on www.google.com (108.177.122.103) at port 443 using host value "www.google.com"


PROBLEM RESULTS:

* None


IGNORED RESULTS:

[--] Expiration validation ignored: 0 expired certificates, 0 expiring certificates

Certificate 1 of 3 (leaf):
        Name: CN=www.google.com
        SANs entries: [www.google.com]
        Issuer: CN=GTS CA 1C3,O=Google Trust Services LLC,C=US
        Serial: 63:4E:19:CA:90:36:79:74:0A:B2:94:4F:E8:A3:E7:C9
        Issued On: 2023-02-08 04:36:30 +0000 UTC
        Expiration: 2023-05-03 04:36:29 +0000 UTC
        Status: [OK] 63d 7h remaining

Certificate 2 of 3 (intermediate):
        Name: CN=GTS CA 1C3,O=Google Trust Services LLC,C=US
        SANs entries: []
        Issuer: CN=GTS Root R1,O=Google Trust Services LLC,C=US
        Serial: 02:03:BC:53:59:6B:34:C7:18:F5:01:50:66
        Issued On: 2020-08-13 00:00:42 +0000 UTC
        Expiration: 2027-09-30 00:00:42 +0000 UTC
        Status: [OK] 1674d 3h remaining

Certificate 3 of 3 (intermediate):
        Name: CN=GTS Root R1,O=Google Trust Services LLC,C=US
        SANs entries: []
        Issuer: CN=GlobalSign Root CA,OU=Root CA,O=GlobalSign nv-sa,C=BE
        Serial: 77:BD:0D:6C:DB:36:F9:1A:EA:21:0F:C4:F0:58:D3:0D
        Issued On: 2020-06-19 00:00:42 +0000 UTC
        Expiration: 2028-01-28 00:00:42 +0000 UTC
        Status: [OK] 1794d 3h remaining

[--] Hostname validation using value "www.google.com" ignored for leaf cert

[--] SANs List validation ignored: 0 SANs entries specified, 1 SANs entries on leaf cert [0 EXPECTED, 0 MISSING, 0 UNEXPECTED]


SUCCESS RESULTS:

* None

 | 'certs_present_intermediate'=2;;;; 'certs_present_leaf'=1;;;; 'certs_present_root'=0;;;; 'certs_present_unknown'=0;;;; 'expires_intermediate'=1674d;30;15;0; 'expires_leaf'=63d;30;15;0; 'time'=43ms;;;;
```
